net/lnet/linux_kernel/klnet_core.c

```diff
@@ -1749,35 +1740,44 @@ M0_INTERNAL int nlx_core_nidstr_encode(struct nlx_core_domain *lcdom,	/* not use
    The returned array must be released using nlx_core_nidstrs_put().
    @param nidary A NULL-terminated (like argv) array of NID strings is returned.
  */
-static int nlx_kcore_nidstrs_get(char * const **nidary)
-{
-	M0_PRE(nlx_kcore_lni_nidstrs != NULL);
-	*nidary = nlx_kcore_lni_nidstrs;
-	m0_atomic64_inc(&nlx_kcore_lni_refcount);
-	return 0;
-}
-
 M0_INTERNAL int nlx_core_nidstrs_get(struct nlx_core_domain *lcdom,
-				     char *const **nidary)
-{
-	return nlx_kcore_nidstrs_get(nidary);
-}
-
-/**
-   Kernel helper to release the strings returned by nlx_kcore_nidstrs_get().
- */
-M0_INTERNAL void nlx_kcore_nidstrs_put(char *const **nidary)
+				     char ***nidary)
 {
-	M0_PRE(*nidary == nlx_kcore_lni_nidstrs);
-	M0_PRE(m0_atomic64_get(&nlx_kcore_lni_refcount) > 0);
-	m0_atomic64_dec(&nlx_kcore_lni_refcount);
-	*nidary = NULL;
+	lnet_process_id_t  id;
+	char              *nidstr;
+	int                i;
+	int                j;
+	int                nr;
+	int                rc;
+
+	for (nr = 0, rc = 0; rc != -ENOENT; ++nr)
+		rc = LNetGetId(nr, &id);
+	M0_ALLOC_ARR(*nidary, nr);
+	if (*nidary == NULL)
+		return M0_ERR(-ENOMEM);
+	for (i = 0; i < nr - 1; ++i) {
+		rc = LNetGetId(i, &id);
+		M0_ASSERT(rc == 0);
+		nidstr = libcfs_nid2str(id.nid);
+		M0_ASSERT(nidstr != NULL);
+		(*nidary)[i] = m0_strdup(nidstr);
+		if ((*nidary)[i] == NULL) {
+			for (j = 0; j < i; ++j)
+				m0_free((*nidary)[j]);
+			return M0_ERR(-ENOMEM);
+		}
+	}
+	return M0_RC(0);
 }
 
 M0_INTERNAL void nlx_core_nidstrs_put(struct nlx_core_domain *lcdom,
-				      char *const **nidary)
+				      char ***nidary)
 {
-	nlx_kcore_nidstrs_put(nidary);
+	int i;
+
+	for (i = 0; (*nidary)[i] != NULL; ++i)
+		m0_free((*nidary)[i]);
+	*nidary = NULL;
```

Comment:
*nidary is allocated in nlx_core_nidstrs_get():
```
1755:	M0_ALLOC_ARR(*nidary, nr);
```
But not freed here.

Is this a memleak?

Reply:
It is. Good catch.